snmp/README.md

```diff
@@ -236,46 +127,314 @@ instances:
-| `<INIT_CONFIG>`        | `{"mibs_folder":"<PATH_TO_ADDITIONAL_MIBS>"}`                                                                                                                                                                                                                                                                          |
-| `<INSTANCE_CONFIG>`    | ```{"ip_address":"%%host%%", "port":"161", "snmp_version":"3", "user":"<USER_NAME>", "authKey":"<PASSWORD>", "privKey":"<PRIVACY_TYPE_KEY>", "authProtocol":"<AUTHENTICATION_PROTOCOL>", "privProtocol":"<PRIVACY_TYPE>", "metrics":[{"MIB":"<MIB_NAME>","symbol":"<SYMBOL>"},{"OID":"<OID>","name":"<OID_NAME>"}]}``` |
-
-
-### Custom metrics
-Metrics collected by the SNMP integration are considered [custom metrics][10], which impacts your [bill][11].
+If necessary, additional metrics can be defined in the instances. These metrics are collected in addition to those in the profile.
+
+#### Metric definition by profile
+
+Profiles can be used interchangeably, such that devices that share MIB dependencies can reuse the same profiles. For example, the Cisco c3850 profile can be used across many Cisco switches.
+
+**Note**: Profiles require a local version of the MIBs to be referenced.
+
+* [Generic router](#generic-router-profile)
+* [F5 Big IP](#f5-big-ip-profile)
+* [Dell iDRAC](#dell-idrac-profile)
+* [Cisco Nexus](#cisco-nexus-profile)
+* [Cisco c3850](#cisco-c3850-device-profile)
+* [Cisco Meraki](#cisco-meraki-profile)
+
+##### Generic router profile
+
+The [generic router profile][7] collects the following metrics.
+
+MIBs needed for local reference: IF-MIB, IP-MIB, TCP-MIB, UDP-MIB.
+
+| Metric                                 | Description                                                                                                                                                                                                                                                               | Tags                     |
+|----------------------------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|--------------------------|
+| `snmp.ifInDiscards`                    | The number of inbound packets chosen to be discarded even though no errors had been detected to prevent them being deliverable to a higher-layer protocol.                                                                                                                | `interface`              |
```

Comment:
These lists are huge! Are you using scripts to generate them? Is there anything we can do to make sure these stay in sync if/when we update profiles, or to make life easier for documentation?

Reply:
Could we use the `metadata.csv` file for this? Adding the profile it's related to in the corresponding metrics description. This will allow updating the metric description within the DD app.

Reply:
I'm fine with metadata.csv, this was a temp solution just so we could ensure that we had metric definitions somewhere...  That being said, I personally am not sure of a good way to show all of the possible metrics you _could_ get from the profiles. The metadata file is usually just a giant list, and I do think that there should be some way to define what metrics you get from each profile

Reply:
FWIW I've been doing these by hand and they are time consuming, but I think basically just as time consuming as writing the metadata file can be

Reply:
As suggested we could add the profile in the metric description with a disclaimer on top of the metric table in the doc to explain that metrics are profile dependant. I can take care of this if i have your approval @naaltman

Reply:
Are you thinking like as a column in the csv, a reference for the profile that the metrics come from? That works for me... There's a decent amount of overlap so that could be good I think

Reply:
Any new ad-hoc `profiles` column in the CSV would need to contain a list of profiles though, as I think metrics in `metadata.csv` must be unique and there may be (and in practice there *are*) multiple profiles associated to a given metric.

The source of truth is the profiles files themselves. Ideally we'd generate `metadata.csv` from those files… and then any rendered tables from those files. (@l0k0ms I think there's some tooling in progress for metadata.csv -> Markdown tables, right?).

Not sure how industrial the profiles -> `metadata.csv` tooling needs to be, but it could be a simple utility script stored in the integration folder. Or even better: a `$ ddev meta snmp ...` command. E.g. we already have `$ ddev meta snmp translate-profile`, so we could add `$ ddev meta snmp generate-metadata`. 🤔

Reply:
ha i like this approach!

There is a logic in the documentation build that indeed inlines a markdown table based on the metadata.csv file (for instance the [Apache integration](https://docs.datadoghq.com/integrations/apache/#metrics). There is also a logic in-app to display the metadata.csv file content automatically in a ["metrics" tab on an integrations tile](https://app.datadoghq.com/account/settings#integrations/apache)

Sadly, we can't change the columns of the `metadata.csv` file as there are too much dependencies for it.

The idea would be to prefix a metrics description with the associated profile:

```
| `snmp.ifInDiscards`                    | The number of inbound packets chosen to be discarded even though no errors had been detected to prevent them being deliverable to a higher-layer protocol. |  `interface`              |
```

Would become: 

```
snmp.ifInDiscards,gauge,,,, [Generic router profile] The number of inbound packets chosen to be discarded even though no errors had been detected to prevent them being deliverable to a higher-layer protocol.,0,snmp,snmp_ifindiscards
```

Two issues here:

* It's not possible to display the list of tags attached to a metric
* We need a type to associate to the metric, are they all `gauge` type ?

Reply:
> There is a logic in the documentation build that indeed inlines a markdown table based on the metadata.csv

I assume that logic is executed at runtime, correct? The Apache docs page shows the rendered Markdown table, but our `README` contains a sentence:

```markdown
See [metadata.csv][9] for a list of metrics provided by this check.
```

I don't think we *need* to put the rendered table in the source of this README (providing it on the docs app via this runtime rendering mechanism is enough, as most users will look up the docs there).

So maybe there's no need for any more tooling. Let's just move stuff to `metadata.csv` as a follow-up — which I assume is what you had in mind @l0k0ms?

> We need a type to associate to the metric, are they all gauge type ?

Not all of them, no. E.g. the `cisco-nexus.yaml` profile as well as the iDRAC one contain metrics converted to `monotonic_count` metrics. (That's the case for `ifInDiscards` for example, which represents a total number of discards of input packets.)

Reply:
indeed you're right, it's done at runtime. And i agree with you we don't "need" too, having it in the metadata.csv is enough.. 

Noted for the metrics type !

Let's merge this and move the metrics table in another PR for sanity